lib/ukrand/random.c

```diff
@@ -0,0 +1,103 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/random.h>
+#include <string.h>
+
+static int uk_hwrand_init(struct uk_init_ctx *ictx __unused)
+{
+	return ukarch_random_init();
+}
+uk_early_initcall(uk_hwrand_init, 0x0);
+
+int uk_random_u32(__u32 *val)
+{
+	int rc = 0;
+
+#if CONFIG_LIBUKRAND_HW
+	rc = ukarch_random_u32(val);
+#elif CONFIG_LIBUKRAND_CHACHA
+	*val = uk_chacha_randr_u32();
+#else /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+#error "Invalid RNG source"
+#endif /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+
+	return rc;
+}
+
+int uk_random_u64(__u64 *val)
+{
+	int rc = 0;
+
+#if CONFIG_LIBUKRAND_HW
+	rc = ukarch_random_u64(val);
+#elif CONFIG_LIBUKRAND_CHACHA
+	*val = uk_chacha_randr_u64();
+#else /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+#error "Invalid RNG source"
+#endif /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+
+	return rc;
+}
+
+int uk_random_seed_u32(__u32 *val)
+{
+	int rc = 0;
+
+#if CONFIG_LIBUKRAND_HW
+	rc = ukarch_random_seed_u32(val);
+#elif CONFIG_LIBUKRAND_CHACHA
+	*val = uk_chacha_randr_u32();
+#else /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+#error "Invalid RNG source"
+#endif /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+
+	return rc;
+}
+
+int uk_random_seed_u64(__u64 *val)
+{
+	int rc = 0;
+
+#if CONFIG_LIBUKRAND_HW
+	rc = ukarch_random_seed_u64(val);
+#elif CONFIG_LIBUKRAND_CHACHA
+	*val = uk_chacha_randr_u64();
+#else /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+#error "Invalid RNG source"
+#endif /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
+
+	return rc;
+}
+
+__ssz uk_random_fill_buffer(void *buf, __sz buflen)
+{
+	__sz step, chunk_size, i;
+	__u64 rd;
+	int ret;
+
+	step = sizeof(__u64);
+	chunk_size = buflen % step;
+
```

Comment:
I think this case is covered, I'll leave an example to be sure we are on the same page:
```
// Assume buflen = 3
step = sizeof(__u64); // 8
chunk_size = buflen % step; // 3 % 8 = 3

// buflen = chunk_size = 3, so the for loop condition is not met (i < buflen - chunk_size)

// this will be true
if (chunk_size > 0) {
  ...
  memcpy((char *)buf + i, &rd, chunk_size); // memcpy((char *)buf + 0, &rd, 3);
  
// buflen = chunk_size = 3 as expected
return buflen;
```
Correct me if I am wrong.